Support torch.bool in torch.sort + CUDA

xw285cornell opened the pull request:
Summary: This might be out-dated, so I'm adding it back and see if we pass all the tests. I'm pretty sure cuda12 is ok.

Test Plan: CI

Differential Revision: D65282650

xw285cornell commented:
@pytorchbot label "topic: not user facing"

xw285cornell commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -18477,11 +18480,18 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
            supports_forward_ad=True,
            supports_fwgrad_bwgrad=True,
            skips=(
+                DecorateInfo(
```
Need to add this because the non-standard bools (uint8) will matter for the sort order (at least for the indices). Wondering if there are some better ideas

eqy replied:
I'm OK with this if the behavior is determined by e.g., cub and we do not care to put in a workaround

xw285cornell commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3320,7 +3320,10 @@ def large_1d_unique():
         # default schema without stable sort
-        yield SampleInput(small_3d_unique(), dim, descending)
+        if not (dtype == torch.bool and torch.device(device).type == 'cuda'):
+            # bool and cuda requires stable sort for stable results, at least
```
adding this because we need stable sort (otherwise it'll fail inductor run which runs on the compiled graph and eager mode and produce different indices). We have the stable sort below and they run fine

nikitaved replied:
The samples seem to be coming from `small_3d_unique`. Isn't it stable in this case?

xw285cornell replied:
small_3d_unique only gives dim and descending but no stable flat. L3344-3346 explicitly set stable=True.

eqy replied:
Would it be possible to account for this in the test cases rather than eliminating some combinations of inputs?

xw285cornell reviewed (commented):
@ngimel @eqy wondering if you can take a look and see if those changes are ok

ngimel reviewed (approved):
lgtm if the tests agree

xw285cornell committed:
Support torch.bool in torch.sort + CUDA (#139409)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/139409

This might be out-dated, so I'm adding it back and see if we pass all the tests. I'm pretty sure cuda12 is ok.

Test Plan: CI

Reviewed By: zou3519, ngimel

Differential Revision: D65282650
